Function Call Graph v2

cameel opened the pull request:
This PR replaces #10332.

I fixed the problems I listed in https://github.com/ethereum/solidity/pull/10332#issuecomment-777617557 and I think I have implemented all the changes that actually affect what's in the graph. ~I'm still not done with some refactors suggested in the old PR like using a queue when visiting nodes adjusting some names or changing the way checks are done. I'm pushing what I have now as a draft but these changes are pretty small so I think I'll be done. The PR should at least already be correct and pass all tests.~ The PR is now ready.

Marenz committed:
Implement a Function Call Graph

Marenz committed:
Add IRGenerator::verifyCallGraphs()

Marenz committed:
CompilerStack: Build call graphs in the analysis phase

Marenz committed:
IRNames::function(): Return correct name if the function is a constructor

Marenz committed:
MultiUseYulFunctionCollector::requestedFunctions(): Refactor the loop to use destructuring

cameel committed:
FunctionCallGraph: operator<< for printing nodes in an unambiguous way

cameel committed:
Test suite for FunctionCallGraph

cameel committed:
CompilerStack: Fix style on local variable name

cameel committed:
CompilerStack: Typo in a comment

cameel committed:
FunctionType: Add assertions against missing type annotations

- This should make it easier to realize that one of the analysis phases has not been executed.

cameel commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,337 @@
+	auto const* functionType = dynamic_cast<FunctionType const*>(_functionCall.expression().annotation().type);
+	solAssert(functionType, "");
+
+	// TODO: Is this actually guaranteed?
+	solAssert(functionType->kind() != FunctionType::Kind::Declaration, "");
```
This was originally asked in https://github.com/ethereum/solidity/pull/10332#discussion_r555941129 and https://github.com/ethereum/solidity/pull/10332#discussion_r555943083.

Honestly, I'm not sure what `FunctionType::Kind::Declaration` would correspond to in the source. I thought that it might be a call to a function that has no definition but I have those in my tests and they do not trigger this assertion. For now I just added this assertion but I can take a closer look at it if needed.

chriseth replied:
Isn't it used for `A.f()` when `A` is a base contract?

cameel replied:
Apparently not. I have that in multiple tests (see for example `inherited_functions_virtual_and_super`) and this assertion is never triggered.

cameel replied:
I managed to trigger it. It's things like `S(42)` (where `S` is a `struct`) or `address(0x0)`. Probably also `C(...)` where `C` is a `contract` but it's hard to verify that without using an address.

Turns out I actually do have quite a few tests cases that do this but this is handled correctly by the check for `FunctionCallKind::FunctionCall` that you suggested earlier so the assert was not failing.

cameel replied:
Assertion removed.

cameel commented on libsolidity/codegen/ir/IRGenerator.cpp:
```diff
@@ -36,18 +36,66 @@
+{
+	using Node = FunctionCallGraphBuilder::Node;
+
+	set<Node const*> reachableNodes = BreadthFirstSearch<Node const*>{{&_root}}.run(
+		[&_graph](Node const* _node, auto&& _addChild) {
+			for (Node const& _child: valueOrDefault(_graph.edges, *_node, {}))
+				_addChild(&_child);
+		}
+	).visited;
```
BFS here is actually a bit superfluous because the graph never contains anything that would not be reachable from an external function or a constructor. We could just as well iterage through the map and collect all non-special nodes.

So the question is: do we want the graph to visit also unreachable stuff? For example all internal and free function definitions?

chriseth replied:
The graph by definition is connected, right? So I think it suffices to return the keys.

cameel replied:
Well, it could be a disconnected graph. But it depends more on how we plan to use it than on the definition so if you say it's fine right now, I'm leaving it at that :)

chriseth replied:
How can it be disconnected? If it is disconnected, then it is a bug.

cameel replied:
I meant that the mathematical graph definition includes both connected and disconnected graphs.

Also, our function call graph could easily contain isolated subgraphs if we wanted it to - e.g. if we made it visit all the internal functions in a contract. Whether it's a bug depends only on whether this is what we want it to do or not.

cameel commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,337 @@
+	// TODO: Is this actually guaranteed?
+	solAssert(functionType->kind() != FunctionType::Kind::Declaration, "");
+
+	if (functionType->kind() == FunctionType::Kind::Internal && !_functionCall.expression().annotation().calledDirectly)
+		// If it's not a direct call, we don't really know which function will be called (it may even
+		// change at runtime). All we can do is to add an edge to the dispatch which in turn has
+		// edges to all functions could possibly be called.
+		add(m_currentNode, SpecialNode::InternalDispatch);
```
Back when this function was visiting events, I noticed that `requiredLookup` annotation always says `Virtual` for identifiers representing the event type in emit statement even though AFAIK events can't be virtual. The weird thing is that it's `Static` for member access. So `emit E()` requires virtual lookup but `emit C.E()` does not - even if it's the same event, just referenced differently.

Is this a bug or a feature? :)

cameel replied:
We talked about it with @chriseth and it's like that because it does not really matter for events so no one cares about what `requiredLookup` says.

I think I'll submit a small PR later to change this but this is not too important.

cameel commented on libsolidity/codegen/ir/IRGenerator.h:
```diff
@@ -115,6 +122,9 @@ class IRGenerator
 	langutil::EVMVersion const m_evmVersion;
 	OptimiserSettings const m_optimiserSettings;
 
+	std::set<FunctionDefinition const*> m_creationFunctionList;
+	std::set<FunctionDefinition const*> m_deployedFunctionList;
```
https://github.com/ethereum/solidity/pull/10332#discussion_r555923381

> Wouldn't it be better to keep generated functions in the IRGenerationContext and retrieve the set from there?

I wanted to change this but one problem is that `generate()` resets the context when it switches from generating construction code to generating deployment code. It would be possible to preserve it but having `resetContext()` not reset the context fully does not seem right to me.

But apart from that - yeah, I think that having it in the context would make this nicer. Currently `IRGenerator` fills the list and then destroys it if you call `verifyGraphs()`. But if you don't call it, the list just stays there and could be reused (I added assertions against that).

cameel replied:
Discussion also here: https://github.com/ethereum/solidity/pull/10973#discussion_r578334715

cameel commented on test/libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,1989 @@
+			modifier m() { emit EvC(1); _; }
+		}
+
+		contract D is C {
+			event EvD1(uint);
+			event EvD2(uint);
+
+			function ext() m external { emit D.EvD1(1); emit EvC(f()); inr(); }
+			function inr() m internal { emit EvD1(1); emit C.EvC(f()); L.inr(); free(); EvD2; }
+
+			function f() internal returns (uint) {}
+		}
```
Note: events that are referenced by name but not emitted (`EvD2` in this example) do not get collected in `emittedEvents`.

And it of course does not include events that are declared in the contract but never emitted.

That's what we want, right?

chriseth replied:
Yes, that's what we want.

cameel commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,337 @@
+	}
+}
+
+void FunctionCallGraphBuilder::enqueueCallable(CallableDeclaration const& _callable)
+{
+	if (!m_graph.edges.count(&_callable))
+	{
+		m_visitQueue.push_back(&_callable);
+
+		// Insert the callable to the graph (with no edges coming out of it) to mark it as visited.
+		m_graph.edges.insert({Node(&_callable), {}});
+	}
+}
```
Note: To avoid visiting functions multiple times I'm inserting keys with empty edge sets into the graph. This also fixes the case where previous implementation would enter an infinite recursion on a function that refers to itself (see test case called `function_self_reference`).

These keys are no longer needed when the graph is ready. Maybe I should add a step that removes them before I return the graph?

chriseth replied:
This way, we can easily get all reachable functions, so I think it is a good feature. Can you document it in the header?

cameel replied:
Documented.

In that case I have also removed the BFS (https://github.com/ethereum/solidity/pull/10973#discussion_r578033659) because it's enough to go through the keys now.

cameel commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,337 @@
+
+		auto typeToString = [](auto const& _var) -> string { return _var->type()->toString(true); };
+		vector<string> parameters = callableDeclaration->parameters() | views::transform(typeToString) | to<vector<string>>();
+		string joinedParameters = parameters | views::join(',') | to<string>();
```
ranges-v3 was giving me weird problems here.

First, it only works when it's split into variables like that. I'm not sure why but I could not get this to compile:
```c++
callableDeclaration->parameters() | views::transform(typeToString) | to<vector<string>>() | views::join(',') | to<string>();
```

Moreover, I also tried to just keep a vector and parameters and join it in the `<<` calls but I was getting some stray `\nul` characters inserted at join points into the string. I was just doing `<< parameters | views::join(",") | to<string>()`. Weirdly, adding `std::` to the string seemed to fix the problem on some input sometimes. The behavior was weirdly inconsistent and I eventually decided that I have already spent too much time trying to solve it. I gave up and used the above.

Generally I find that getting the pipes to work in expressions is sometimes a pain and requires lots of experimenting. Errors are often not very informative and just point at the `|` and say that types do not match.

cameel commented on test/libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,1989 @@
+			{"function D.ext1()", "function D.inr1()"},
+			{"function D.ext1()", "function C.inr2()"},
+		}},
+	};
```
Note: interface functions get an edge from `Entry` too even though they do not have implementations. Same for unimplemented internal functions in abstract contracts (like `C.inr1()` in the example above). That's what we want, right?

chriseth replied:
I think it's fine.

cameel commented on test/libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,1989 @@
+			function q() m1 internal virtual { L.f(); }
+		}
+
+		contract D is C {
+			modifier m2() { q(); _; new C(); }
+
+			function p() m2 internal { C.q(); }
+			function q() m2 internal override virtual {}
+		}
+
+		contract E is D {
+			modifier m1() override { _; }
+			modifier m3() { p(); _; }
+
+			constructor() D() m1 E.m3 {}
+			function ext() external m1 E.m3 { inr(); }
+			function inr() internal m1 E.m3 { L.f(); }
+
+			function q() internal override {}
+		}
```
Note: the current implementation detects `E` as the contract creating `C`. Not `D`, even though the modifier that does this is itself defined in `D` and also only applied to functions defined in `D`.

Do we want to change this?

chriseth replied:
If you compile `D` in isolation then it does not create `C`, so I think it's fine.

cameel commented on test/libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,1989 @@
+			function g() internal {}
+		}
+
+		contract C {
+			modifier m1() virtual { _; }
+
+			function q() m1 internal virtual { L.f(); }
+		}
+
+		contract D is C {
+			modifier m2() { q(); _; new C(); }
+
+			function p() m2 internal { C.q(); }
+			function q() m2 internal override virtual {}
+		}
```
Note: currently a modifier is included in the graph as something called by the function it modifies. For example in the above we have `p() -> m2`, `p() -> C.q()`, `m2 -> q()` and `C.q() -> m1` (ignoring other edges).

But the actual execution order is different. It's more like `m2 -> p() -> m1 -> C.q()` because the modifier wraps function body. Should I change that?

cameel replied:
I talked about it with @chriseth.

My suggestion of putting the modifier in between the call and the function has the downside that it would look like the modifier is calling all the functions that use it which would be misleading (there are separate inlined instances of the modifier that do not have anything to do with each other). If we wanted to do this, it would be better to just "inline" the modifier in the graph too - making it look like the function is doing all the stuff the modifier does.

@chriseth suggested that the way it's done currently is probably fine though. After giving it some thought I think I agree. It does have an effect similar to "inlining" the modifier in the graph without actually hiding it.

cameel commented on test/libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,1989 @@
+		contract F is D(fFD()), E {
+			uint f = fVarF();
+
+			constructor (uint) E(fFE()) {}
+		}
+	)"s);
+	tuple<CallGraphMap, CallGraphMap> graphs = collectGraphs(*compilerStack);
+
+	map<string, EdgeNames> expectedCreationEdges = {
+		{"C", {
+			{"Entry", "constructor of C"},
+			{"Entry", "function fVarC()"},
+		}},
+		{"D", {
+			{"Entry", "constructor of C"},
+			{"Entry", "function fVarC()"},
+			{"constructor of C", "constructor of D"},
+			{"constructor of C", "function fD()"},
+			{"constructor of C", "function fVarD()"},
+		}},
+		{"E", {
+			{"Entry", "constructor of C"},
+			{"Entry", "function fVarC()"},
+			{"constructor of C", "constructor of E"},
+			{"constructor of C", "function fVarE()"},
+		}},
+		{"F", {
+			{"Entry", "constructor of C"},
+			{"Entry", "function fVarC()"},
+			{"constructor of C", "constructor of D"},
+			{"constructor of C", "function fD()"},
+			{"constructor of C", "function fVarD()"},
+			{"constructor of D", "constructor of E"},
+			{"constructor of D", "function fVarE()"},
+			{"constructor of E", "constructor of F"},
+			{"constructor of E", "function fFD()"},
+			{"constructor of E", "function fVarF()"},
+			{"constructor of F", "function fFE()"},
+		}},
+	};
```
I'm not sure how the order in which constructors call each other (and functions called from state variable initializers, inheritance specifiers and constructor modifier list) should look like. This probably needs some discussion so I'll only point out some things that stand out here:
- Functions called from state variable initializers currently are marked as called by the constructor one level below (or `Entry` when we're at the bottom of the hierarchy).
- Functions evaluated to get parameters for the base constructor end up having a different edge depending on whether they're on inheritance specifier list or modifier list. Note for example that in `F` we have `constructor of E -> fFD` and `constructor of F -> fFE()`. They look like called from different contracts but they're both called for constructor parameters in `F`.
    - This might actually be what @chriseth meant in https://github.com/ethereum/solidity/pull/10332#pullrequestreview-588767032. Changing `m_currentNode` there could change this.

cameel replied:
This has now changed so I'm marking it as resolved. See https://github.com/ethereum/solidity/pull/10973#discussion_r579244040 for further discussion.

BTW, the test case visible in the snippet above has been renamed to `inheritance_specifiers` and I removed the one that was previously under this name. I think that `diamond_inheritance` covers everything from the old test case and does it better.

cameel commented on libsolidity/analysis/FunctionCallGraph.h:
```diff
@@ -0,0 +1,120 @@
+namespace solidity::frontend
+{
+
+/**
+ * Creates a Function call graph for a contract at the granularity of Solidity
+ * functions and modifiers. The graph can represent the situation either at contract creation
+ * or at deployment time.
+ *
+ * Includes the following special nodes:
+ *  - Entry: represents a call from the outside of the contract.
+ *    At deployment this is the node that connects to all the functions exposed through the external
+ *    interface. At contract creation it connects to the constructor and variable initializers in
+ *    the bottom-most contract in the linearized inheritance order.
+ *  - InternalDispatch: Represents the internal dispatch function, which calls internal functions
+ *    determined at runtime by values of variables and expressions. Functions that are not called
+ *    right away get an edge from this node.
+ *
+ *  Nodes are a variant of either the enum SpecialNode or a CallableDeclaration which currently
+ *  can be a function or a modifier. There are no nodes representing event calls. Instead all
+ *  emitted events and created contracts are gathered in separate sets included in the graph just
+ *  for that purpose.
+ *
+ *  Auto-generated getter functions for public state variables are ignored.
+ *
+ *  Only calls reachable from an Entry node are included in the graph.
+ */
```
This originally mentioned that there could be nodes other than callables - e.g. contracts to represent implicit constructors. Or that function calls without a definition could be represented by `FunctionCall` nodes.

None of that was implemented so I removed that from the docstring. I have also changed `Node` to hold `CallableDeclaration` rather than `ASTNode`. If we want to add more node types, I'd rather just add them as separate variants and not have an overly generic pointer that has to be always cast and asserted against.

Anyway, do we want any of that in this PR or it it fine the way it is now?

chriseth replied:
Looks good!

cameel reviewed (commented):
Since the PR is pretty big, below I'm pointing out some places where I'd like some discussion/decision/explanation.

bshastry commented on test/CMakeLists.txt:
```diff
@@ -97,6 +97,7 @@ set(libsolidity_sources
     libsolidity/SyntaxTest.cpp
     libsolidity/SyntaxTest.h
     libsolidity/ViewPureChecker.cpp
+	libsolidity/analysis/FunctionCallGraph.cpp
```
Tab indenting here. Others are white space indented.

cameel replied:
Good catch. How did I miss that :)

We should really decide on a single indentation style for CMake files :)

chriseth commented on libsolidity/analysis/FunctionCallGraph.h:
```diff
@@ -0,0 +1,120 @@
+/**
+ * Creates a Function call graph for a contract at the granularity of Solidity
+ * functions and modifiers. The graph can represent the situation either at contract creation
+ * or at deployment time.
```
```suggestion
 * or after deployment.
```

chriseth commented on libsolidity/analysis/FunctionCallGraph.h:
```diff
@@ -0,0 +1,120 @@
+ *
+ * Includes the following special nodes:
+ *  - Entry: represents a call from the outside of the contract.
+ *    At deployment this is the node that connects to all the functions exposed through the external
```
```suggestion
 *    After deployment this is the node that connects to all the functions exposed through the external
```

bshastry commented on libsolidity/analysis/FunctionCallGraph.h:
```diff
@@ -0,0 +1,120 @@
+class FunctionCallGraphBuilder: private ASTConstVisitor
+{
+public:
+	enum class SpecialNode {
```
Shouldn't the brace start in a newline?

chriseth replied:
```suggestion
	enum class SpecialNode
	{
```
(or everything on a single line)

chriseth commented on libsolidity/analysis/FunctionCallGraph.h:
```diff
@@ -0,0 +1,120 @@
+ * Creates a Function call graph for a contract at the granularity of Solidity
+ * functions and modifiers. The graph can represent the situation either at contract creation
+ * or at deployment time.
+ *
```
Please mention that the edges are not ordered, i.e. the graph does not contain any temporal relations.

bshastry commented on libsolidity/analysis/FunctionCallGraph.h:
```diff
@@ -0,0 +1,120 @@
+		/// Contract for which this is the graph
+		ContractDefinition const& contract;
+
+		std::map<Node, std::set<Node, CompareByID>, CompareByID> edges;
```
Could you please add a comment to briefly describe what this data structure represents like the other struct members?

bshastry replied:
I suppose this member holds (directed?) edges of the callgraph?

cameel replied:
Yes. Docstring added.

chriseth commented on libsolidity/analysis/FunctionCallGraph.h:
```diff
@@ -0,0 +1,120 @@
+ *  emitted events and created contracts are gathered in separate sets included in the graph just
+ *  for that purpose.
+ *
+ *  Auto-generated getter functions for public state variables are ignored.
```
```suggestion
 *  Auto-generated getter functions for public state variables are ignored, but function calls inside initial assignments are included in the creation graph.
```

bshastry commented on libsolidity/ast/Types.cpp:
```diff
@@ -2632,11 +2632,13 @@ FunctionType::FunctionType(FunctionDefinition const& _function, Kind _kind):
 
 	for (ASTPointer<VariableDeclaration> const& var: _function.parameters())
 	{
+		solAssert(var->annotation().type, "Parameter type is not yet available in the AST.");
```
For what function parameter types is this assertion expected to fail?

cameel replied:
It will fail if `type` is null, which may be the case if you forget to run `DeclarationTypeChecker` during analysis. Without this assertion you will just get a memory violation at a random place and only if you happen to use the annotation. I had a really hard time debugging this and such an assertion would have saved me a lot of time.

It's not all that relevant to this PR anymore because I switched to creating the whole `CompilerStack` in tests (rather than running specific analysis phases manually). There are a few commits with small tweaks like this one in this PR. I could extract them into a separate PR but I wasn't sure if it's worth it since these tweaks are so small (some of them are just typo fixes).

chriseth commented on libsolidity/analysis/FunctionCallGraph.h:
```diff
@@ -0,0 +1,120 @@
+	};
+
+	static std::unique_ptr<ContractCallGraph> buildCreationGraph(ContractDefinition const& _contract);
+	static std::unique_ptr<ContractCallGraph> buildDeploymentGraph(
```
DeployedGraph?

cameel replied:
Done. I changed it here and in all other names this PR introduces.

chriseth commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,337 @@
+
+bool FunctionCallGraphBuilder::CompareByID::operator()(Node const& _lhs, int64_t _rhs) const
+{
+	solAssert(!holds_alternative<SpecialNode>(_lhs), "");
```
Why can we assemu that?

cameel replied:
I see it more as a precondition than an assumption.

When called from `operator(Node, Node)`, this will always be true. When called directly, it's up to the caller to ensure this is satisfied.

chriseth commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,337 @@
+unique_ptr<FunctionCallGraphBuilder::ContractCallGraph> FunctionCallGraphBuilder::buildCreationGraph(ContractDefinition const& _contract)
+{
+	FunctionCallGraphBuilder builder(_contract);
+
```
Should we set the current node here just to be safe?

cameel replied:
Added an assert to enforce that.

bshastry commented on libsolidity/codegen/MultiUseYulFunctionCollector.cpp:
```diff
@@ -32,11 +32,11 @@ using namespace solidity::frontend;
+	for (auto const& [name, code]: m_requestedFunctions)
 	{
-		solAssert(f.second != "<<STUB<<", "");
+		solAssert(code != "<<STUB<<", "");
```
Does the compiler warn that `name` is unused here? I've had this question for some time myself :-)

cameel replied:
Good point. It does not. Which is weird because I remember having problems with unused variables in destructuring.

chriseth commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,337 @@
+		{
+			builder.functionReferenced(*base->constructor());
+
+			// Constructors and functions called in state variable initializers have an edge either from
```
I would be fine with simplifying this to having an edge from entry for all of them, since none of these calls are explicit, or does that create a problem?

cameel replied:
I don't think it'll be a problem. We're just losing some detail but if we do not need that detail then it does not mater.

I'll change that.

cameel replied:
Done.

Now it works like this:
- Constructors always get an edge from `Entry`.
- Functions called in initial assignments to stage variables always get an edge from `Entry`.
- Functions called on inheritance specifier list get an edge from the the constructor of the current contract. If that contract has no constructor then the edge is from `Entry`.
    - Before it was always an edge from the constructor of the previous contract in linearized hierachy.
    - I think that it might make sense for this always have an edge from `Entry` but then it would be inconsistent with the edges that calls on the modifier list get (as pointed out in https://github.com/ethereum/solidity/pull/10973#discussion_r578056366).
    - I think that my question about modifier edges is relevant here too: https://github.com/ethereum/solidity/pull/10973#discussion_r578050881

You can see the effect of this change in the `inheritance_specifiers` test case.

chriseth commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,337 @@
+	builder.m_currentNode = SpecialNode::Entry;
+	builder.processQueue();
+
+	return make_unique<FunctionCallGraphBuilder::ContractCallGraph>(std::move(builder.m_graph));
```
Why not just return by value?

cameel replied:
Well, I thought the same thing and I even changed it but then I reverted back because `CompilerStack` was storing it in a map and I wanted to avoid unnecessary copies.

But, now that I think about it again, I could go back to that version. We actually do the same thing with `Contract` struct in `CompilerStack` so there there should be no problems as long as I add `move()`s at the right places.

cameel replied:
Done.

chriseth commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,337 @@
+		auto const* variable = dynamic_cast<VariableDeclaration const*>(&functionType->declaration());
+
+		if (function)
+			builder.functionReferenced(*function);
```
Do we need virtual lookup here?

cameel replied:
I think we don't. I guess it wouldn't break anything if we did but it's not needed.

This loop goes over the functions returned by `interfaceFunctionList()` of the contract we're building the graph for. If you override an external function, `interfaceFunctionList()` will encounter the overridden version first and then will ignore all the functions from base contracts that it overrides because they have the same signature. So we don't unless you can change the signature when overriding.

I added this situation to one of the test cases and it works as expected - virtual functions from the base classes do not show up in the graph if the current contract overrides them.

chriseth commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,337 @@
+			solAssert(variable, "");
+	}
+
+	if (_contract.fallbackFunction())
```
Do we need to check if the base has a fallback or receive function?

cameel replied:
I have a test case for this and fallback/receive functions from base classes get added as expected.

cameel replied:
I just realized that they can be virtual too so I added a test case for that. It works as expected.

chriseth commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,337 @@
+	for (Node const& dispatchTarget: valueOrDefault(_creationGraph.edges, SpecialNode::InternalDispatch, {}))
+	{
+		solAssert(!holds_alternative<SpecialNode>(dispatchTarget), "");
+		solAssert(get<CallableDeclaration const*>(dispatchTarget) != nullptr, "");
```
Isn't this redundant with the next?

cameel replied:
Yeah. Looks like I just changed `ASTNode` to `CallableDeclaration` here but I didn't stop long enough to notice that this makes the whole line redundant :) Removed.

chriseth commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,337 @@
+	{
+		solAssert(!holds_alternative<SpecialNode>(dispatchTarget), "");
+		solAssert(get<CallableDeclaration const*>(dispatchTarget) != nullptr, "");
+		solAssert(dynamic_cast<CallableDeclaration const*>(get<CallableDeclaration const*>(dispatchTarget)) != nullptr, "");
```
Why the dynamic cast?

cameel replied:
The line used to look like this:
```c++
solAssert(dynamic_cast<CallableDeclaration const*>(get<ASTNode const*>(dispatchTarget)) != nullptr, "");
```

Now that I replaced `ASTNode` with `CallableDeclaration` in `Node`, this became redundant.

I'm removing this line.

bshastry reviewed (commented):
First pass of mostly minor comments. Will take a closer look in a bit.

chriseth commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,337 @@
+
+bool FunctionCallGraphBuilder::visit(FunctionCall const& _functionCall)
+{
+	solAssert(holds_alternative<SpecialNode>(m_currentNode) || get<CallableDeclaration const*>(m_currentNode) != nullptr, "");
```
Shouldn't this always be the case? If you really want to check this all the time, maybe change the type to a `std::reference_wrapper`?

cameel replied:
It should never be null if the code is correct but that's what assertions are for :)

Maybe `reference_wrapper` would be better than a pointer but I don't really want to change it everywhere at this point so I'll remove the assert. In this particular place I'm not dereferencing the pointer anyway, it was more of a "just in case" situation.

chriseth commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,337 @@
+		// change at runtime). All we can do is to add an edge to the dispatch which in turn has
+		// edges to all functions could possibly be called.
+		add(m_currentNode, SpecialNode::InternalDispatch);
+
```
Why does this not add an edge if the function is called directly?

chriseth replied:
Ah, that is done in `functionReferenced`... Wouldn't it be more natural to only add edges here?

cameel replied:
Currently `visit(FunctionCall)` finds indirect calls **to** functions while `visit(Identifier)` and `visit(MemberAccess)` find direct and indirect calls **from** functions.

If I changed this, `visit(FunctionCall)` would be addin4 indirect calls **to** and direct calls **from** functions and I would still need to keep `visit(Identifier)` and `visit(MemberAccess)` to detect indirect calls **from** functions.

Both make sense and in either case I can't get rid of any of these functions.

chriseth commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,337 @@
+	auto const* functionType = dynamic_cast<FunctionType const*>(_emitStatement.eventCall().expression().annotation().type);
+	solAssert(functionType, "");
+
+	m_graph.emittedEvents.insert(dynamic_cast<EventDefinition const*>(&functionType->declaration()));
```
```suggestion
	m_graph.emittedEvents.insert(&dynamic_cast<EventDefinition const&>(functionType->declaration()));
```

cameel replied:
Good point. This will give me `bad_cast` instead of just passing `nullptr` to the function when it does not inherit from `EventDefinition`.

bshastry commented on test/libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,1989 @@
+		{"D", {
+			{"Entry", "constructor of C"},
+			{"constructor of C", "constructor of D"},
+			{"constructor of C", "function free()"},
```
Not sure I understood the second entry in the creation call graph here. During the creation of `D`, `C` is created and `free` is called. However, why does `constructor of C` point to `constructor of D`?

cameel replied:
Is that the `inheritance_specifiers` test case?

The constructor of C will run before constructor of D so it does not necessarily make sense the other way.

Well, maybe it should just be an edge from `Entry` after all. That's what @chriseth actually said in https://github.com/ethereum/solidity/pull/10973#discussion_r578300322

Please also take a look at https://github.com/ethereum/solidity/pull/10973#discussion_r578056366 which is relevant here.

chriseth commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,337 @@
+			if (auto const contractType = dynamic_cast<ContractType const*>(typeType->actualType()))
+			{
+				solAssert(contractType->isSuper(), "");
+				functionDef =
+					&functionDef->resolveVirtual(
+						m_graph.contract,
+						contractType->contractDefinition().superContract(m_graph.contract)
+					);
```
```suggestion
				functionDef = &functionDef->resolveVirtual(
					m_graph.contract,
					contractType->contractDefinition().superContract(m_graph.contract)
				);
```

chriseth commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,337 @@
+		solAssert(*_memberAccess.annotation().requiredLookup == VirtualLookup::Static, "");
+
+	functionReferenced(*functionDef, _memberAccess.annotation().calledDirectly);
+	return;
```
```suggestion
```

chriseth commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,337 @@
+	return true;
+}
+
+void FunctionCallGraphBuilder::endVisit(MemberAccess const& _memberAccess)
```
Don't we also need 'visit' here?

chriseth replied:
If not, then it would be easier to change all the others to `endVisit`, too.

cameel replied:
I changed it to `visit()` and it does pass all my tests so I guess either works. Done.

chriseth commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,337 @@
+
+void FunctionCallGraphBuilder::endVisit(ModifierInvocation const& _modifierInvocation)
+{
+	VirtualLookup const& requiredLookup = *_modifierInvocation.name().annotation().requiredLookup;
```
Can move inside the if.

chriseth commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,337 @@
+	{
+		m_currentNode = m_visitQueue.front();
+		m_visitQueue.pop_front();
+		get<CallableDeclaration const*>(m_currentNode)->accept(*this);
```
Does this work?
```suggestion
		get<CallableDeclaration const&>(*m_currentNode).accept(*this);
```

cameel replied:
Nope.
```
error: no match for ‘operator*’ (operand type is ‘solidity::frontend::FunctionCallGraphBuilder::Node’
```

I guess it would work with `dynamic_cast` but for a variant you need the exact type.

But thanks for pointing this out - `accept()` changes `m_currentNode` so this actually should have an assert that it's not a special node.

chriseth commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,337 @@
+	m_currentNode = SpecialNode::Entry;
+}
+
+bool FunctionCallGraphBuilder::add(Node _caller, Node _callee)
```
Is the return value used?

cameel replied:
It's not. Removed.

chriseth commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,337 @@
+		switch (specialNode)
+		{
+			case SpecialNode::InternalDispatch:
+				_out<< "InternalDispatch";
```
```suggestion
				_out << "InternalDispatch";
```

chriseth commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,337 @@
+
+	if (holds_alternative<SpecialNode>(_node))
+	{
+		auto specialNode = get<SpecialNode>(_node);
```
No need for this local variable

chriseth commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,337 @@
+				_out<< "InternalDispatch";
+				break;
+			case SpecialNode::Entry:
+				_out<< "Entry";
```
```suggestion
				_out << "Entry";
```

chriseth commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,337 @@
+		solAssert(callableDeclaration, "");
+
+		auto const* function = dynamic_cast<FunctionDefinition const *>(callableDeclaration);
+		auto const* event = dynamic_cast<EventDefinition const *>(callableDeclaration);
```
Can this still happen?

cameel replied:
Yes. Not when printing the graph but I am using this operator in tests to print out events from `emittedEvents`.

Also, technically `Node` can store any `CallableDeclaration`. There's nothing stopping anyone from putting an event in `Node`.

chriseth commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,337 @@
+
+		auto typeToString = [](auto const& _var) -> string { return _var->type()->toString(true); };
+		vector<string> parameters = callableDeclaration->parameters() | views::transform(typeToString) | to<vector<string>>();
+		string joinedParameters = parameters | views::join(',') | to<string>();
```
You could use `joinHumanReadable`

cameel replied:
I can't chain `joinHumanReadable()` with `|` though :)

But I'll switch to `joinHumanReadable()` because I had problems with chaining it anyway (see https://github.com/ethereum/solidity/pull/10973#discussion_r578044154).

cameel replied:
Done.

chriseth commented on libsolidity/codegen/ir/IRGenerator.cpp:
```diff
@@ -180,11 +257,20 @@ string IRGenerator::generate(Block const& _block)
 }
 
-void IRGenerator::generateQueuedFunctions()
+set<FunctionDefinition const*> IRGenerator::generateQueuedFunctions()
```
Does it make sense to store this information in the context instead?

chriseth replied:
Ah this is also discussed here: https://github.com/ethereum/solidity/pull/10973/files#r578037647

cameel replied:
Yeah. In short, I think it would if not for the fact that the context gets reset during `generate()`.

I'll mark this one as resolved to keep discussion under the other comment.

chriseth commented:
This PR contains a core logic of 200 lines, yet there is a test file of ten times that - is that really necessary, especially since we cross-check the result at each compilation?

bshastry commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,337 @@
+	// All functions present in internal dispatch at creation time could potentially be pointers
+	// assigned to state variables and as such may be reachable at deployment time as well.
+	builder.m_currentNode = SpecialNode::InternalDispatch;
+	for (Node const& dispatchTarget: valueOrDefault(_creationGraph.edges, SpecialNode::InternalDispatch, {}))
```
The default value passed to `valueOrDefault` is an rvalue reference but return is a const reference. ASan tells me that this can lead to stack-use-after-scope. I interpret this to be an rvalue reference that has already gone out of scope to be bound to a const ref.

I'm not sure what is a good solution here.

cameel replied:
Oh. Well, what if I use a local variable to store the empty set? Check out the updated code.

This does make this pattern considerably uglier though.

cameel commented:
I just pushed updated code. I fixed/changed everything that is now marked as resolved. Of the unresolved ones, the only two on my todo list are: returning the graph by value (https://github.com/ethereum/solidity/pull/10973#discussion_r578300886) and simplifying the construction-time edges (https://github.com/ethereum/solidity/pull/10973#discussion_r578300322). The others are waiting for feedback.

> This PR contains a core logic of 200 lines, yet there is a test file of ten times that - is that really necessary, especially since we cross-check the result at each compilation?

It's more like 500 lines of implementation vs 2000 lines of tests so not *that* bad :) Well, what can I say, long and repetitive tests are long and repetitive. 2k lines is a lot and it also was a pain to write but I don't really think it's too much for this PR. The implementation is relatively short but I think this feature has quite a big "surface". Just look at this PR. There are ~90 comments even before anyone started seriously reviewing these tests. The old PR ran into ~250 and we had to close it to ease it's suffering. While a lot of them are just style issues there was also an unusually high number that pointed out actual bugs. I think this is because there are just a lot of different combinations in which things can call each other and it's hard to visualize that without actually going through them all systematically. Which is exactly what I did here and the size of these tests just shows how big that surface is.

I don't think the cross-check is sufficient for testing this. It ignores constructors, modifiers, events and contract creation. And there were actually bugs or missing pieces of implementation in all of these things. It only checks which functions are present but not if they're connected correctly. The check also runs only on semantic tests (because graph verification is in IR generator) and I'm not sure if they cover all the important cases for this.

After writing these tests I do feel pretty confident in just going ahead, changing stuff because I know it's covered. And if it's not, I add a new case and it stays fixed instead of getting broken again while I work on the next fix. I don't think I would feel that confident about it if I was relying only on these checks. I might just have gotten bogged down in the implementation, fixing one thing while breaking another two.

One way I could have made these tests smaller would be by just finding existing .sol files, loading them from files at runtime and writing expectations for them. I think it would take me a lot more time that just writing this out though. So in a way I traded test size for development time here :)

cameel commented:
This is now ready for another round. I squashed the fixups because there were a lot of them and it's probably easier just to go through the implementation again.

bshastry commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,347 @@
+
+	if (holds_alternative<SpecialNode>(_lhs))
+		return get<SpecialNode>(_lhs) < get<SpecialNode>(_rhs);
+	return get<CallableDeclaration const*>(_lhs)->id() < get<CallableDeclaration const*>(_rhs)->id();
```
Could we use solutil's generic visitor on this variant?

cameel replied:
We could, but I'm not sure it's worth it on a simple variant with just two cases. I also don't think it would make it much more readable.

chriseth commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,347 @@
+
+		// Functions called from the inheritance specifier should have an edge from the constructor
+		// for consistency with functions called from constructor modifiers.
+		for (auto const& inheritanceSpecifier: base->baseContracts())
```
Base contracts specified as modifiers in derived constructors are handled in those constructors, right?

cameel replied:
Yeah. They get processed by `builder.functionReferenced(*base->constructor())` below.

chriseth commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,347 @@
+	// assigned to state variables and as such may be reachable after deployment as well.
+	builder.m_currentNode = SpecialNode::InternalDispatch;
+	set<Node, CompareByID> emptySet;
+	for (Node const& dispatchTarget: valueOrDefault(_creationGraph.edges, SpecialNode::InternalDispatch, emptySet))
```
Why not?
```suggestion
	for (Node const& dispatchTarget: valueOrDefault(_creationGraph.edges, SpecialNode::InternalDispatch, set<Node, CompareByID>{}))
```
This has the advantage that the variable cannot be referenced later on.

cameel replied:
It does seem to work for me (even though it returns a temporary) and was my preferred solution but [@bshastry reported that asan complains about `stack-use-after-scope`](https://github.com/ethereum/solidity/pull/10973#discussion_r578339963).

chriseth replied:
That is weird - isn't its lifetime extended to the end of the range-based for loop?

cameel replied:
Might be a false-positive due to the fact that I am passing `dispatchTarget` into a function which does actually stored it. It's safe only because the set is empty and asan might not be accounting for that.

Should I just use the temporary and ignore the potential warning? The old version wasn't failing in CI so it would work.

There are also two other uses of `valueOrDefault()` which @bshastry did not point out so I guess there were no warnings for them. I changed them along with this one but if we think it's safe, I could change back at least those two places.

cameel replied:
I have just pushed two fixups that restore the temporaries.

Oh, and it was actually one other place that uses `valueOrDefault()`, not two. Looks like removing the BFS eliminated the one in `IRGenerator.cpp`.

chriseth commented on libsolidity/codegen/ir/IRGenerator.cpp:
```diff
@@ -76,11 +119,31 @@ pair<string, string> IRGenerator::run(
 {
+	// Remember to call verifyCallGraphs() (which clears the list of generated functions) if you
+	// want to reuse the generator.
+	solAssert(m_creationFunctionList.empty(), "");
```
I'm pretty sure these can be local variables in this function - can't they?

cameel replied:
They only reason they're not is that they're used by `verifyCallGraphs()` which gets called separately by `CompilerStack`.

We could make them local if we made `generate()` call `verifyCallGraphs()` automatically. We would then also have to pass the graphs from `CompilerStack` either via `generate()` or `IRGenerator` constructor or the context.

chriseth replied:
If they are in the annotations, then we can just call it here, can't we? It's not too elegant to call that function from the compilerstack anyway, I think.

cameel replied:
Sure. But as I said, I'll be making a separate PR to change them to annotations. In that case I'll also change this to use locals that that other PR.

chriseth replied:
ok!

cameel replied:
Done in #10990.

chriseth commented:
Please squash so it can be merged.

bshastry commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,343 @@
+			case SpecialNode::Entry:
+				_out << "Entry";
+				break;
+			default: solAssert(false, "Invalid SpecialNode type");
```
```suggestion
			default:
				solAssert(false, "Invalid SpecialNode type");
```

cameel replied:
In my opinion it looks fine but I see we don't usually put it on the same line so I'll change it.

cameel replied:
Fixed in #10995.

bshastry commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,343 @@
+	if (holds_alternative<SpecialNode>(_node))
+		switch (get<SpecialNode>(_node))
+		{
+			case SpecialNode::InternalDispatch:
```
case statements need one fewer indentation according to our style guide.

cameel replied:
Right. Good point.

cameel replied:
Fixed in #10995.

cameel commented:
Squashed.

I have also added a small change in the builder docstring to account for the fact that after my changes it's not only the bottom-most constructor that has an edge from `Entry`.

bshastry commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,343 @@
+	// All functions present in internal dispatch at creation time could potentially be pointers
+	// assigned to state variables and as such may be reachable after deployment as well.
+	builder.m_currentNode = SpecialNode::InternalDispatch;
+	for (Node const& dispatchTarget: valueOrDefault(_creationGraph.edges, SpecialNode::InternalDispatch, {}))
```
iiuc life-time extension of dispatchTarget happens until the loop lives. But is the rvalue temporary `{}` live until it is bound to the range variable?

cameel replied:
Why wouldn't it be? I might be wrong but I think this is just a false-positive (https://github.com/ethereum/solidity/pull/10973#discussion_r580320382).

BTW, do you get warnings about both uses of `valueOrDefault()` or just this one?

bshastry replied:
I will take a closer look and post an update.

bshastry replied:
Update posted here: https://gist.github.com/bshastry/16bf461806a56fd2bcb8499a4e7b0151

cameel replied:
Well, I'll need to read up a bit on the rules for lifetime extension for temporaries. For now I just found an SO question on a similar topic that indicates that this indeed might be a problem ([Is it safe to use a C++11 range-based for-loop with an rvalue range-init?](https://stackoverflow.com/questions/30448182/is-it-safe-to-use-a-c11-range-based-for-loop-with-an-rvalue-range-init)). The example cited there is safe but in our case I think that this condition for lifetime extension is not satisfied:

> The temporary to which the reference is bound or the temporary that is the complete object of a subobject to which the reference is bound persists for the lifetime of the reference except:
> - (...)
> - A temporary bound to a reference parameter in a function call (...)

I remember that @ekpyron had to try hard to even make this function work and I'm not sure if he expected that the case where the default is not a simple value would be relevant :) I wonder if the function could be reworked to account for that...

bshastry replied:
Yeah, that function needs to be debugged for complex objects like sets of structs or something to evaluate lifetime of temporary. Temporary fix would be to bind the default param to a value on the stack e.g., local variable of the same type.
